### DISCUSSION TITLE
Don't Understand Patrick IneractionTest script

### DISCUSSION THREAD

**Discussion by EngrPips** [Cyfrin/foundry-full-course-cu#129] [Q&A] [answered]
Patrick was typing the code in the InetractionTest.t.sol file 

```solidity
function testUserCanFundInteractions() public {
        FundFundMe fundFundMe = new FundFundMe();
        vm.prank(USER);
        vm.deal(USER, 1e18);
        fundFundMe.fundFundMe(address(fundFundMe));

        address funder = fundMe.getFunder(0);
        assertEq(funder, USER);
    }
```
and all of a sudden the next thing i saw was the code below, i can't seem to understand what is going. been trying to understand since yesterday but all my effort have been futile
```solidity
function testUserCanFundInteractions() public {
        FundFundMe fundFundMe = new FundFundMe();
        fundFundMe.fundFundMe(address(fundFundMe));

        WithdrawFundMe withdrawFundMe = new WithdrawFundMe();
        withdrawFundMe.withdrawFundMe(address(fundFundMe));
        assert(address(fundMe).balance == 0);
    }
```
timestamp of the particular slide in the course video is around ```2:08:53```
i would appreciate any help at all @PatrickAlphaC

**Comment by alymurtazamemon** [comment, marked as answer]
@EngrPips In this test;
```solidity
function testUserCanFundInteractions() public {
    FundFundMe fundFundMe = new FundFundMe();
    vm.prank(USER);
    vm.deal(USER, 1e18);
    fundFundMe.fundFundMe(address(fundFundMe));

    address funder = fundMe.getFunder(0);
    assertEq(funder, USER);
}
```
He is first initializing a contract which is a script called `FundFundMe` then using its `fundFundMe` function he just funded the contract. After that he checked the funder should be `USER` because before funding we used `vm.prank(USER);` which call the next call on behalf of the address provided in the `prank` function.

In this test;
```solidity
function testUserCanFundInteractions() public {
  FundFundMe fundFundMe = new FundFundMe();
  fundFundMe.fundFundMe(address(fundFundMe));

  WithdrawFundMe withdrawFundMe = new WithdrawFundMe();
  withdrawFundMe.withdrawFundMe(address(fundFundMe));
  assert(address(fundMe).balance == 0);
}
```
same as previous test he created the `FundFundMe` contract script and funded it but this time the test contract is the funder because we did not use any `prank` function.

Next similar to `FundFundMe` we have another script inside the `interactions` file `WithdrawFundMe` which has a function `withdrawFundMe` which calls the `withdraw` function of `FundMe` contract.

So he just withdraw all fo the funds and checked the remaining balance should be zero.

I hope this will make sense!

**Reply by cromewar** [reply]
Marking this as answer.

**Comment by EngrPips** [comment]
@alymurtazamemon  i am trying to make it make sense by trying to wrap my head around it , firstly that means they are two different tests and he later erased one and leave just  this test below 
```solidity 
function testUserCanFundInteractions() public {
  FundFundMe fundFundMe = new FundFundMe();
  fundFundMe.fundFundMe(address(fundFundMe));

  WithdrawFundMe withdrawFundMe = new WithdrawFundMe();
  withdrawFundMe.withdrawFundMe(address(fundFundMe));
  assert(address(fundMe).balance == 0);
}
```
and secondly does a foundry contract/script automatically have balance or where did the contract get the funds to fund FundMe because i didn't see us send any ether to the contract during the whole testing process

**Reply by alymurtazamemon** [reply]
@EngrPips Yes, in the video first he tested first then then convert it to this test.

Second, if you see the fund me script you will see that we are passing the value and above it we are using the `startBroadcast` which will use the default address to get the funds.

**Comment by EngrPips** [comment]
I notice we called the fundFundMe function on the FundFundMe script and not the run function, Obviously our fundFundMe function did not call the run function anywhere in the script, So is it that whenever we call any function on our script it would first re-route the call to the run function ??

@alymurtazamemon

**Reply by alymurtazamemon** [reply]
@EngrPips No, it will not re-route these are just like separate functions and Patrick created them separately intentionally so we can use the `fundFundMe` and `withdrawFundMe` by providing the address of the contract.

**Reply by alymurtazamemon** [reply]
@EngrPips Basically the difference is, if we call the `run` function then using the `foundryDevops` plugin it will automatically get the recently deployed address from the `broadcast` folder but for the separate function we manually need to give the address. So in the tests, we wanted to give out our own addresses due to that we created a separate function.

**Reply by EngrPips** [reply]
Makes sense to me now, so we ended up not using the run function because we wanted to test both functions individually, one more thing that comes to my mind now is can we instead call the run function and that would call the fund function instead of us manually calling the fund function ?
@alymurtazamemon

**Reply by alymurtazamemon** [reply]
@EngrPips Yes we can do but we do not want this, because then it will take the other contract address from broadcast folder.

**Comment by EngrPips** [comment]
Also in this our InteractionTest.t.sol file we already seem to use the latest  address of the FundMe contract to make our call to the Interactions.s.sol file , because  we did our deployment with DeployFundMe contract which would deploy a new FundMe i believe.
```solidity
//SPDX-License-Identifier: MIT

pragma solidity ^ 0.8.18;

import {Test, console} from "forge-std/Test.sol";
import {FundMe} from "../../src/FundMe.sol";
import {DeployFundMe} from "../../script/DeployFundMe.s.sol";
import {FundFundMe, WithdrawFundMe} from "../../script/Interactions.s.sol";


contract InteractionTest is Test {

    FundMe public fundMe;

    address USER = makeAddr("user");
    uint256 constant SEND_VALUE = 0.1 ether;
    uint256 constant STARTING_BALANCE = 100 ether;
    uint256 constant GAS_PRICE = 1;
    function setUp() external {
        DeployFundMe deploy = new DeployFundMe();
        fundMe = deploy.run();
        vm.deal(USER, STARTING_BALANCE);
    }

    // function testUserCanFundInteractions() public {
    //     FundFundMe fundFundMe = new FundFundMe();
    //     vm.prank(USER);
    //     vm.deal(USER, 1e18);
    //     fundFundMe.fundFundMe(address(fundFundMe));

    //     address funder = fundMe.getFunder(0);
    //     assertEq(funder, USER);
    // }

    function testUserCanFundInteractions() public {
        FundFundMe fundFundMe = new FundFundMe();
        fundFundMe.fundFundMe(address(fundMe));

        WithdrawFundMe withdrawFundMe = new WithdrawFundMe();
        withdrawFundMe.withdrawFundMe(address(fundMe));
        assert(address(fundMe).balance == 0);
    }
}
```
so we called this function below
 ```solidity 
function fundFundMe(address mostRecentlyDeployed) public {}
```
 with the address of the new created FundMe contract already so would the call execute directly or it would instead trigger the run function 

## The basis of my question  is if we make  a call to a function inside of a script which is not the run function does that function  get triggered or it would instead trigger the run function. @PatrickAlphaC  @alymurtazamemon

**Reply by alymurtazamemon** [reply]
@EngrPips Answered it above.

**Reply by EngrPips** [reply]
so what was the use case of the DevOpsTools we imported if the run function is not called at all,  did he just showed us that so that we know that it exist? 

because now it seem like we make the funding and withdrawing call directly to the funding and withdrawing function in the FundFundMe contract and WithdrawFundMe contract using the address of the FundMe contract deployed by our Deployer script during the setUp of InteractionsTest

**Reply by alymurtazamemon** [reply]
@EngrPips DevOpsTools help us to get the latest deployed contract address from the broadcast folder, we do not need to worry about the network we are running on, i.e anvil, sepolia, or mainnet. It will automatically pick the right address for us depending on the network.

**Comment by EngrPips** [comment]
I have learned a ton from this conversation everything seem clear to me now, thank you so much @alymurtazamemon  , i was stuck on this chapter since yesterday lol. now i owe you a follow on github. hahahha

**Comment by EngrPips** [comment]
followed you on twitter as well, thank you so much
@alymurtazamemon
